### PR TITLE
Rosstalk Listener

### PR DESCRIPTION
**Pull request by onfire4g05**
This allows Companion to listen to RossTalk commands and will recall buttons in Companion. This will resolve issue #474.

The help in the config explains, but basically:

`CC 4:6`

Will press/depress `4.6` in Companion.

### CODE REVIEW

**Comment by krocheck** [comment]
I'm OK with this.  It handles press/release and it is off by default.  Appears to follow RossTalk specifications.  Thoughts @haakonnessjoen or @Julusian?

**Comment by haakonnessjoen** [comment]
Maybe our other services aren’t better. But there seems to be a lack of error handling here. (Looking with my mobile phone :p. please arrest me if I am wrong)

That should be added first. But nice idea :)

File: lib/rosstalk.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+				try {
+					this.listen();
+				} catch(e) {
+					console.log('Error listening for artnet', e);
+				}
+			}
+		});
+
+		this.system.on('set_userconfig_key', (key, val) => {
+			if(key !== 'rosstalk_enabled') return;
```
Comment (krocheck) [inline lib/rosstalk.js]: This should have curly braces

File: lib/rosstalk.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+			this.config['rosstalk_enabled'] = val;
+
+			try {
+				if(val === true) this.listen();
```
Comment (krocheck) [inline lib/rosstalk.js]: Ditto

File: lib/rosstalk.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+	process_incomming(data) {
+		// Type, bank/page, CC/bnt number
+		const match = data.match(/(CC) ([0-9]*)\:([0-9]*)/);
+		if(match === null) {
+			this.system.emit('log', 'rosstalk', 'warn', `Invalid incomming RossTalk command: ${data}`);
+			return;
+		}
+
+		if(match[1] === 'CC') this.press_button(match[2], match[3]);
```
Comment (krocheck) [inline lib/rosstalk.js]: Ditto

File: lib/rosstalk.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+		this.system.emit('bank_pressed', bank, button, true);
+
+		setTimeout(() => {
+			this.system.emit('bank_pressed', bank, button, false);
+			this.system.emit('log', 'rosstalk', 'info', `Release button ${bank}.${button}`);
+		}, this.release_time);
+	}
+
+	listen() {
+		if(this.active) return; // already listening
```
Comment (krocheck) [inline lib/rosstalk.js]: Ditto

File: lib/rosstalk.js (hunk 1 above)
Comment (haakonnessjoen) [reply lib/rosstalk.js]: There should be a space after the «if» codeword also

File: lib/rosstalk.js [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+ *
+ */
+
+let net = require('net');
+
+class rosstalk {
+	constructor(_system) {
+		this.system = _system;
+		this.active = false;
+		this.socket = this.createSocket();
```
Comment (krocheck) [inline lib/rosstalk.js]: Should this have a try/catch?

Comment (onfire4g05) [reply lib/rosstalk.js]: It didn't look like `createServer` threw exceptions. It looks like errors are passed to the `error` event. I did just add some error logging, which will then close to port, if there's a problem creating the socket or while it's open. I'm not sure if more is needed.

**Pull request merged**